File: ios/src/main/kotlin/org.jetbrains.kotlinconf/data/FavoritesManager.kt
package org.jetbrains.kotlinconf.data

import kotlinx.cinterop.*
import org.jetbrains.kotlinconf.ui.*
import org.jetbrains.kotlinconf.util.*
import platform.Foundation.*
import platform.CoreData.*

class FavoritesManager {
    fun isFavorite(session: Session) = getFavorite(session) != null

    private fun getFavorite(session: Session): Favorite? {
        return AppContext.localFavorites.firstOrNull { it.sessionId == session.id }
    }

    fun toggleFavorite(
        session: Session,
        onComplete: (Boolean) -> Unit
    ) {
        val newFavorite = !isFavorite(session)
        val service = KonfService({ log(it.localizedDescription) })

        setLocalFavorite(session, isFavorite = newFavorite)
        onComplete(newFavorite)

        val uuid = appDelegate.userUuid
        if (newFavorite) {
            service.addFavorite(session, uuid, onComplete = {})
        } else {
            service.deleteFavorite(session, uuid, onComplete = {})
        }
    }

    private fun setLocalFavorite(session: Session, isFavorite: Boolean) {
        val favorite = getFavorite(session)
        if (favorite != null && !isFavorite) {
            AppContext.localFavorites.remove(favorite)
        } else if (isFavorite) {
            AppContext.localFavorites.add(Favorite(session.id))
        }
    }

    fun getFavoriteSessionIds(): List<String> {
        return AppContext.localFavorites.map { it.sessionId }.filterNotNull()
    }
}

File: konf-swift/konfswift/data/FavoritesManager.swift
import Foundation

class FavoritesManager {
    private static let KEY = "favorites"
    private let userDefaults = UserDefaults.standard

    func isFavorite(session: KSession) -> Bool {
        return getFavorite(session: session) != nil
    }

    private func getFavorite(session: KSession) -> KFavorite? {
        let moc = AppDelegate.me.managedObjectContext

        let request = NSMakeFetchRequest(for: KFavorite.self)
        request.predicate = NSPredicate(format: "sessionId == %@", session.id!)
        request.fetchLimit = 1

        guard let favorite = (try? moc.fetch(request))?.first else {
            return nil
        }

        return favorite
    }

    func toggleFavorite(
        for session: KSession,
        errorHandler: @escaping (Error) -> (),
        onComplete: @escaping (Bool) -> ()
    ) {
        let newFavorite = !isFavorite(session: session)

        let service = KonfService(errorHandler: errorHandler)

        let completionHandler: () -> () = {
            assert(Thread.isMainThread)

            do {
                try self.setLocalFavorite(for: session, isFavorite: newFavorite)
                onComplete(newFavorite)
            } catch let e {
                errorHandler(e)
            }
        }

        let uuid = AppDelegate.me.userUuid
        if (newFavorite) {
            service.addFavorite(session, uuid: uuid, onComplete: completionHandler)
        } else {
            service.deleteFavorite(session, uuid: uuid, onComplete: completionHandler)
        }
    }

    func setLocalFavorite(for session: KSession, isFavorite: Bool) throws {
        let moc = AppDelegate.me.managedObjectContext

        if let favorite = getFavorite(session: session) {
            if (!isFavorite) {
                moc.delete(favorite)
            }
        } else if isFavorite {
            let favorite = NSEntityDescription.insertNewObject(
                forEntityName: String(describing: KFavorite.self), into: moc) as! KFavorite
            favorite.sessionId = session.id

            try moc.save()
        }
    }

    func getFavoriteItemIds() -> [String] {
        let moc = AppDelegate.me.managedObjectContext

        let favorites: [KFavorite] = (try? moc.fetch(NSMakeFetchRequest(for: KFavorite.self))) ?? []
        return favorites.flatMap { $0.sessionId }
    }
}
